crates/sui-types/src/transaction.rs

```diff
@@ -2460,47 +2641,63 @@ impl InputObjects {
-            .map(|(_, obj)| SharedInput::Existing(obj.compute_object_reference()))
-            .chain(self.deleted.iter().map(|info| SharedInput::Deleted(*info)))
+            .filter(|obj| obj.is_shared_object())
+            .map(|obj| {
+                obj.to_shared_input()
+                    .expect("already filtered for shared objects")
+            })
             .collect()
     }
 
     pub fn transaction_dependencies(&self) -> BTreeSet<TransactionDigest> {
-        let mut dependencies: BTreeSet<TransactionDigest> = self
-            .objects
+        self.objects
             .iter()
-            .map(|(_, obj)| obj.previous_transaction)
-            .collect();
-
-        for (_, _, _, digest) in &self.deleted {
-            dependencies.insert(*digest);
-        }
-
-        dependencies
+            .map(|obj| obj.get_previous_transaction())
+            .collect()
     }
 
     pub fn mutable_inputs(&self) -> BTreeMap<ObjectID, (VersionDigest, Owner)> {
         self.objects
             .iter()
-            .filter_map(|(kind, object)| match kind {
-                InputObjectKind::MovePackage(_) => None,
-                InputObjectKind::ImmOrOwnedMoveObject(object_ref) => {
-                    if object.is_immutable() {
-                        None
-                    } else {
-                        Some((object_ref.0, ((object_ref.1, object_ref.2), object.owner)))
+            .filter_map(
+                |ObjectReadResult {
+                     input_object_kind,
+                     object,
+                 }| match (input_object_kind, object) {
+                    (InputObjectKind::MovePackage(_), _) => None,
+                    (
+                        InputObjectKind::ImmOrOwnedMoveObject(object_ref),
+                        ObjectReadResultKind::Object(object),
+                    ) => {
+                        if object.is_immutable() {
+                            None
+                        } else {
+                            Some((object_ref.0, ((object_ref.1, object_ref.2), object.owner)))
+                        }
                     }
-                }
-                InputObjectKind::SharedMoveObject { mutable, .. } => {
-                    if *mutable {
-                        let oref = object.compute_object_reference();
-                        Some((oref.0, ((oref.1, oref.2), object.owner)))
-                    } else {
-                        None
+                    (
+                        InputObjectKind::ImmOrOwnedMoveObject(_),
+                        ObjectReadResultKind::DeletedSharedObject(_, _),
+                    ) => {
+                        unreachable!()
                     }
-                }
-            })
+                    (
+                        InputObjectKind::SharedMoveObject { .. },
+                        ObjectReadResultKind::DeletedSharedObject(_, _),
+                    ) => None,
```

Comment:
Need to think a bit more/look at where this is used a bit more before I feel confident in saying one way or the other, but this doesn't feel quite right to me as we will treat deleted shared objects in different ways depending on whether or not it shows up as mutable or not, so saying they're not a mutable input seems a bit odd maybe?

Reply:
I'm just trying to follow the existing code as closely as possible. Deleted shared objects were previously not part of this, because they were not in `self.objects`

Reply:
the tests definitely fail spectacularly if i try to include mutable deleted inputs here